src/test/groovy/graphql/Issue2068.groovy

```diff
@@ -0,0 +1,264 @@
+package graphql
+
+import graphql.execution.instrumentation.dataloader.DataLoaderDispatcherInstrumentation
+import graphql.schema.DataFetcher
+import graphql.schema.DataFetchingEnvironment
+import graphql.schema.StaticDataFetcher
+import graphql.schema.idl.RuntimeWiring
+import org.apache.commons.lang3.concurrent.BasicThreadFactory
+import org.dataloader.BatchLoader
+import org.dataloader.DataLoader
+import org.dataloader.DataLoaderOptions
+import org.dataloader.DataLoaderRegistry
+import spock.lang.Specification
+
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.CompletionStage
+import java.util.concurrent.SynchronousQueue
+import java.util.concurrent.ThreadFactory
+import java.util.concurrent.ThreadPoolExecutor
+import java.util.concurrent.TimeUnit
+
+import static graphql.schema.idl.TypeRuntimeWiring.newTypeWiring
+import static graphql.ExecutionInput.newExecutionInput
+
+class Issue2068 extends Specification {
+    def "deadlock attempt"() {
+        setup:
+        def sdl = """
+        type Nation {
+            name: String
+        }
+        
+        type Toy {
+            name: String
+        }
+        
+        type Owner {
+            nation: Nation
+        }
+        
+        type Cat {
+            toys: [Toy]
+        }
+        
+        type Dog {
+            owner: Owner
+        }
+        
+        type Pets {
+            cats: [Cat]
+            dogs: [Dog]
+        }
+        
+        type Query {
+            pets: Pets
+        }
+        """
+
+        def cats = [['id': "cat-1"]]
+        def dogs = [['id': "dog-1"]]
+
+        ThreadFactory threadFactory = new BasicThreadFactory.Builder()
+                .namingPattern("resolver-chain-thread-%d").build()
+        def executor = new ThreadPoolExecutor(15, 15, 0L,
+                TimeUnit.MILLISECONDS, new SynchronousQueue<>(), threadFactory,
+                new ThreadPoolExecutor.CallerRunsPolicy())
+
+        DataFetcher nationsDf = { env -> env.getDataLoader("owner.nation").load(env) }
+        DataFetcher ownersDf = { env -> env.getDataLoader("dog.owner").load(env) }
+
+        def wiring = RuntimeWiring.newRuntimeWiring()
+                .type(newTypeWiring("Query")
+                        .dataFetcher("pets", new StaticDataFetcher(['cats': cats, 'dogs': dogs])))
+                .type(newTypeWiring("Cat")
+                        .dataFetcher("toys", new StaticDataFetcher(new List<Object>() {
```

Comment:
maybe a `Collections.singletonList` is better here? `size` returns 1, but then `toArrray`, `iterator` returns both null

Reply:
my bad,  just spotted the "iterator" that throws exception!

Reply:
@KammererTob what about: 

```
			 new java.util.AbstractList<>() {

				@Override
				public int size() {
					return 1;
				}

				@Override
				public Object get(int index) {
					throw new RuntimeException("simulated failure");
				}
			};
```

by default, the Iterator will call `get()`, throwing exception, like in your code.

Reply:
Thanks for the feedback. I've adjusted the code as by your comment and i also renamed the whole test block to better reflect what it is testing.